File: backends/analog/QOJulia.jl
using Configurations
using JSON


# @option struct Operator
#     coefficient::Int = 1
#     qreg::Array{Int} = []
#     qmode::Array{Int} = []
#  end
#
#  @option struct Dissipation
#     jumps::Array{Int} = []
#  end

 @option struct AnalogGate
    duration::Int = 1
#     unitary::Array{Operator} = []
#     dissipation::Array{Dissipation} = []
 end

@option struct AnalogCircuit
#     definitions::Array{Int} = []
#     registers::Array{Int} = []
    sequence::Array{AnalogGate} = []
    n_qreg::Int = 0

#     n_qreg::Int = nothing
#     n_qmode::Int = nothing
 end


function test_analog_circuit(circ_json)
    println(AnalogCircuit())
#     println(circ_json)
    d = JSON.parse(circ_json);
    println(d)
#     aa = from_dict(AnalogCircuit, d);
#     println(aa)

#     for op in circ.sequence
#         println(op)
#     end
#
#     d = JSON.json(to_dict(circ))
end

# print(test_analog_circuit())
# println(AnalogGate())

File: backends/analog/qojulia.py
#%%
from rich import print as pprint
import numpy as np
from juliacall import Main as jl
from typing import List

from pydantic import BaseModel


#%%
class AnalogGate(BaseModel):
    duration: int = 1


class AnalogCircuit(BaseModel):
    sequence: List[AnalogGate] = []
    n_qreg: int = None


#%%
c = AnalogCircuit(sequence=[AnalogGate()], n_qreg=1)
pprint(c)

#%%
jl.seval('include("./QOJulia.jl")')

#%%
jl.test_analog_circuit(c.model_dump_json())

#%%
